website: 1.1 blog post (#2897)
electric-sql/electric: website/blog/posts/2025-08-13-electricsql-v1.1-released.md

authors: [icehaunter, balegas]
image: /img/blog/electric-1.1-released/header.png
tags: [ai, sync]
outline: [2, 3]
post: true
---

<script setup>
import StorageComparisonChart from '../../src/components/StorageComparisonChart.vue'
import StorageComparisonChartColumn from '../../src/components/StorageComparisonChartColumn.vue'
import StorageEngineDiagram from '/static/img/blog/electric-v1.1-new-storage/storage-engine-diagram.svg'

// Raw performance data
const readPerformanceData = {
  '1.0.24': [131.68, 154.07, 1690.00, 2550.00],
  '1.1.0': [2.64, 2.10, 14.00, 14.80]
}

const writePerformanceData = {
  '1.0.24': [0.32, 3.49, 167.68, 1.4, 15.5, 712.58],
  '1.1.0': [0.01, 0.12, 1.65, 0.26, 3.39, 99.29]
}

// Compute normalized data (relative to 1.0.24)
const normalizedReadData = {
  '1.0.24': readPerformanceData['1.0.24'].map(() => 1.0),
  '1.1.0': readPerformanceData['1.1.0'].map((value, index) => value / readPerformanceData['1.0.24'][index])
}

const normalizedWriteData = {
  '1.0.24': writePerformanceData['1.0.24'].map(() => 1.0),
  '1.1.0': writePerformanceData['1.1.0'].map((value, index) => value / writePerformanceData['1.0.24'][index])
}
</script>

> Electric is a [Postgres](https://www.postgresql.org/) sync engine that streams database changes to millions of concurrent users in real time. Our mission is simple: be faster than Postgres.

Reviewer: I don't think we have a better link than this on scalability

```suggestion
Electric is a [Postgres](https://www.postgresql.org/) sync engine that [streams database changes to millions of concurrent users in real time](https://electric-sql.com/blog/2024/12/10/electric-beta-release#scalable) by syncing through CDNs. Our mission is simple: be faster than Postgres.
```

Reviewer: We could also have a short section here something like:

```
Add a 3‑bullet “Why this architecture is different” box (near the “How Electric works” section):

- CDN‑native fan‑out: HTTP is cacheable; CDNs de‑duplicate shape requests and collapse load.
- Cost physics: No persistent socket tax; lower egress + fewer hot servers at scale.
- Operational simplicity: Shape log persistence means zero‑downtime deploys without stateful socket fleets.
```

We’ve been working toward that goal for the last year, but when usage started growing and workloads began pushing the limits of the system, we hit a wall — our storage engine had become a bottleneck. Replication lag was creeping up, CPU usage was climbing higher and higher and the system was having trouble keeping pace with Postgres itself. It was time to act, and we went big: we built our own storage engine from the ground up.

> The result? Up to 100X faster writes and 73X faster reads on SSD. This is the story of how we made it.

Reviewer: Let's remove all the "up to" — it makes the results sound imprecise.

```suggestion
The result? 100X faster writes and 73X faster reads on SSD. This is the story of how we made it.
```

Electric's job sounds simple, until you have hundreds of thousands of shapes to match and millions of clients hitting the system at once. At that scale, every microsecond counts. Fall behind Postgres throughput and real-time becomes sluggish and Postgres's [WAL](https://www.postgresql.org/docs/current/wal-intro.html) starts to pile up. It's not pretty.

<figure>
  <a href="/img/api/shape-log.jpg">
    <img srcset="/img/api/shape-log.sm.png 1064w, /img/api/shape-log.png 1396w"
        sizes="(max-width: 767px) 600px, 1396px"
        src="/img/api/shape-log.png"
    alt="Flow diagram showing how changes are matched to shapes, appended to shape logs, and streamed to clients"
    />
  </a>
  <figcaption class="figure-caption text-end">
    Flow of changes into shape logs and out to clients.
  </figcaption>

</figure>

## Storage is the beating heart of Electric

> One [difference](https://expertofobsolescence.substack.com/p/the-hard-things-about-sync) between sync engines and other types of real‑time systems is that sync engines don't miss changes. Real‑time systems typically offer at‑most‑once delivery or temporal buffering windows. If you lose connection, you're done. In Electric, users can resume shapes at any point in history. This makes the sync engine dramatically simpler to use but storage becomes the beating heart of the sytem.

Reviewer: ```suggestion
One [difference](https://expertofobsolescence.substack.com/p/the-hard-things-about-sync) between sync engines and other types of real‑time systems is that sync engines don't miss changes. Real‑time systems typically offer at‑most‑once delivery or temporal buffering windows. If you lose connection, you're done. In Electric, users can resume shapes at any point in history. This makes the sync engine dramatically simpler to use but storage becomes the beating heart of the system.
```

**Finding a chunk**: When a client requests data starting from a specific offset, we do a binary search on the index to locate the appropriate chunk pointer, retrieve that chunk, and scan it from there, as explained above.

**Lock‑free**: Because shape logs are append‑only, offset pointers are always added to the end of the index. This means that the index can be read and written without any locking. We keep a number of chunk pointers in memory to avoid reading the index from disk for [live requests](/docs/api/http#live-mode).

### Read-only mode

The new architecture decouples readers and writers, allowing the Electric server connected to Postgres to share access with other Electric servers accessing the shape logs in read-only mode. This unlocks new capabilities.

**Horizontal read scaling**: Electric is already quite scalable [behind a CDN](/blog/2024/12/10/electric-beta-release), but read-only mode allows horizontal scaling of the read path. Electric can scale to any read workload.

> **Zero‑downtime deployments**: Keeping the system available during deploys is criticial for our Cloud infrastructe. We achieve zero-downtime deployments by allowing servers to continue serving data from shape logs, while the old and new server switchover the connection to Postgres.

Reviewer: ```suggestion
**Zero‑downtime deployments**: Keeping the system available during deploys is critical for our Cloud infrastructure. We achieve zero-downtime deployments by allowing servers to continue serving data from shape logs, while the old and new server switchover the connection to Postgres.
```

## Performance evaluation

Enough talk — show me the numbers! To validate our new storage engine, we ran a series of performance tests.

> Microbenchmarks ran on a local SSD (MacBook Air M4), and end‑to‑end benchmarks ran on AWS t2.medium instances with network-attached storage (EFS).

Reviewer: is the benchmark code public? We should link to it if we can

### Micro-benchmarks

We conducted a series of microbenchmarks to evaluate the new storage engine against CubDB. We saw strong speedups on both SSDs and EFS, with up to 101x and 7x faster writes and 73x and 172x faster reads, respectively.

#### Write performance

This benchmark consists of appending a fixed number of rows to a shape log. With CubDB, every insertion needs to update the index to locate the right chunk to write to. With the new engine, we simply append to the latest chunk and only modify the index when we reach the chunk‑size limit.

> On local SSDs, the new engine achieved up to 101x faster writes when appending 1,000 rows. These results were a bit surprising; we hadn't really seen this before, as appending this many rows to a single shape log is uncommon.

Reviewer: No sure — it's a perfectly reasonable benchmark — I just think we should drop the commentary.

> On local SSDs, the new engine achieved up to 101x faster writes when appending 1,000 rows. These results were a bit surprising; we hadn't really seen this before, as appending this many rows to a single shape log is uncommon.

Reviewer: ```suggestion
On local SSDs, the new engine achieved 101x faster writes when appending 1,000 rows. These results were a bit surprising; we hadn't really seen this before, as appending this many rows to a single shape log is uncommon.
```

    { label: '1.1.0', data: [21.5, 53.5, 72.1, 120.5, 157.9, 199.3, 197.8, 225.1, 239.7, 274.9, 334.6] }
  ]"
  :labels="['5','100','200','300','400','500','600','700','800','900','1000']"
  x-axis-title="Connected clients"
  y-axis-title="Latency"
  y-axis-suffix=" ms"
  :columns="1"
  :height="260"
/>

## Lessons learned

Starting with CubDB was a good choice. It got us to v1.0 with minimal issues, giving us time to ship software that just works. Before addressing storage performance, we've fixed tons of bugs, made Electric a reliable system and learned a great deal about how it behaves under real‑world workloads, surfacing bottlenecks in many other places that were far more important to our core proposition. Had we chosen to build our own storage system from day one, surely we would have made the wrong assumptions and done lots on premature optimizations. Instead, our production experience gave us the insight we needed to design the right solution in the right moment.

This isn't a universal playbook. It worked well for us because the scope of what we had to build was relatively small, and it was one of our team's core areas of expertise.

## What's next

> We've taken a big step toward our ambitious goal: being **faster than Postgres**. The new storage engine has delivered significant performance gains with up to 100x faster writes on SSD.

Reviewer: ```suggestion
We've taken a big step toward our ambitious goal: being **faster than Postgres**. The new storage engine has delivered significant performance gains with 100x faster writes on SSD.
```

import StorageEngineDiagram from '/static/img/blog/electric-v1.1-new-storage/storage-engine-diagram.svg'

// Raw performance data
const readPerformanceData = {
  '1.0.24': [131.68, 154.07, 1690.00, 2550.00],
  '1.1.0': [2.64, 2.10, 14.00, 14.80]
}

const writePerformanceData = {
  '1.0.24': [0.32, 3.49, 167.68, 1.4, 15.5, 712.58],
  '1.1.0': [0.01, 0.12, 1.65, 0.26, 3.39, 99.29]
}

// Compute normalized data (relative to 1.0.24)
const normalizedReadData = {
  '1.0.24': readPerformanceData['1.0.24'].map(() => 1.0),
  '1.1.0': readPerformanceData['1.1.0'].map((value, index) => value / readPerformanceData['1.0.24'][index])
}

const normalizedWriteData = {
  '1.0.24': writePerformanceData['1.0.24'].map(() => 1.0),
  '1.1.0': writePerformanceData['1.1.0'].map((value, index) => value / writePerformanceData['1.0.24'][index])
}
</script>

Electric is a [Postgres](https://www.postgresql.org/) sync engine that [streams database changes to millions of concurrent users in real time](https://electric-sql.com/blog/2024/12/10/electric-beta-release#scalable). Our mission is simple: be faster than Postgres.

> We’ve been working toward that goal for the last year, but when usage started growing and workloads began pushing the limits of the system, we hit a wall — our storage engine had become a bottleneck. Replication lag was creeping up, CPU usage was climbing higher and higher and the system was having trouble keeping pace with Postgres itself. It was time to act, and we went big: we built our own storage engine from the ground up.

Reviewer: Tighten up

```suggestion
After a year of progress, rapid growth exposed our limits. The storage engine had become a critical bottleneck — replication lag climbing, CPUs maxing out, and the system was having trouble keeping pace with the largest deployments. We made a bold decision: build our own storage engine from scratch.
```
